### PR TITLE
multi: update blockchain and mempool error types.

### PR DESCRIPTION
**Pull request by dnldd**
This updates the blockchain and mempool error types to leverage go 1.13 errors.Is/As functionality as well as confirm to
the error infrastructure best practices.

### CODE REVIEW

File: blockchain/error.go [hunk 1]
```diff
@@ -71,574 +70,461 @@ func (e AssertError) Error() string {
...
-// rules.  The caller can use type assertions to determine if a failure was
-// specifically due to a rule violation and access the ErrorCode field to
-// ascertain the specific reason for the rule violation.
+// RuleError identifies a rule violation related to blocks. It has
+// full support for errors.Is and errors.As, so the caller can ascertain the
+// specific reason for the error by checking the underlying error.
 type RuleError struct {
-	ErrorCode   ErrorCode // Describes the kind of error
-	Description string    // Human readable description of the issue
+	Description string
```
Comment (davecgh) [inline blockchain/error.go:513]: Not a big deal, but all the others put the error first.  Better to be consistent imo.

**Review by davecgh — CHANGES_REQUESTED** [review]
Thanks for the PR.  It looks pretty good overall.  There are a couple of incorrect cases I identified inline as well as many opportunities to simplify the code with by making full use of the new infrastructure also identified inline.

File: blockmanager.go [hunk 2]
```diff
@@ -633,7 +633,7 @@ func errToWireRejectCode(err error) (wire.RejectCode, string) {
 	switch {
 	case errors.As(err, &berr):
 		// Convert the chain error to a reject code.
-		switch berr.ErrorCode {
+		switch berr.Err {
```
[the remark this answers is not in the document]
Comment (dnldd) [reply blockmanager.go:636]: I provided a default case for each nested switch.

[the remark this answers is not in the document]
Comment (davecgh) [reply blockmanager.go:681]: This is still not really correct.  Now that both `blockchain` and `mempool` use the new error infrastructure, I'd suggest simplifying all this to just be.

```Go
func errToWireRejectCode(err error) (wire.RejectCode, string) {
	// The default reason to reject a transaction/block is due to it being
	// invalid somehow.
	code := wire.RejectInvalid
	var reason string

	// Convert recognized errors to a reject code.
	switch {
	// Rejected due to duplicate.
	case errors.Is(err, blockchain.ErrDuplicateBlock):
		code = wire.RejectDuplicate
		reason = err.Error()

	// Rejected due to obsolete version.
	case errors.Is(err, blockchain.ErrBlockVersionTooOld):
		code = wire.RejectObsolete
		reason = err.Error()

	// Rejected due to checkpoint.
	case errors.Is .... the rest

	...

	// Error codes which map to a duplicate transaction .....
	case errors.Is(err, mempool.ErrMempoolDoubleSpend),
		errors.Is(err, mempool.ErrAlreadyVoted),
		.... the rest:

		code = wire.RejectCheckpoint
		reason = err.Error()

	default:
		reason = fmt.Sprintf("rejected: %v", err)
	}
	return code, reason
}

In other words, there is no longer a need to manually unwrap and check things.  Just check for the recognized things directly and let `errors.Is` handle the unwrapping.

File: blockmanager.go [hunk 3]
```diff
@@ -1853,7 +1859,7 @@ func isDoubleSpendOrDuplicateError(err error) bool {
 
 	var rerr mempool.TxRuleError
 	if errors.As(merr.Err, &rerr) {
-		switch rerr.ErrorCode {
+		switch rerr.Err {
```
Comment (davecgh) [inline blockmanager.go:1862]: Same deal with all this.

```Go
func isDoubleSpendOrDuplicateError(err error) bool {
	switch {
	case errors.Is(err, mempool.ErrDuplicate):
		return true
	case errors.Is(err, mempool.ErrAlreadyExists):
		return true
	case errors.Is(err, blockchain.ErrMissingTxOut):
		return true
	}

	return false
}
```

File: internal/mempool/error_test.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2014 The btcsuite developers
+// Copyright (c) 2015-2018 The Decred developers
```
Comment (davecgh) [inline internal/mempool/error_test.go:2]: 2020

File: internal/mempool/error_test.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		{ErrTooManyTSpends, "ErrTooManyTSpends"},
+		{ErrTSpendMinedOnAncestor, "ErrTSpendMinedOnAncestor"},
+		{ErrTSpendInvalidExpiry, "ErrTSpendInvalidExpiry"},
+	}
+
+	t.Logf("Running %d tests", len(tests))
+	for i, test := range tests {
+		result := test.in.Error()
+		if result != test.want {
+			t.Errorf("String #%d\n got: %s want: %s", i, result,
```
Comment (davecgh) [inline internal/mempool/error_test.go:47]: Not `String`.
```Go
			t.Errorf("#%d: got: %s want: %s", i, result, test.want)
```

File: internal/mempool/error_test.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+			RuleError{Description: "human-readable error"},
+			"human-readable error",
+		},
+	}
+
+	t.Logf("Running %d tests", len(tests))
+	for i, test := range tests {
+		result := test.in.Error()
+		if result != test.want {
+			t.Errorf("Error #%d\n got: %s want: %s", i, result,
```
Comment (davecgh) [inline internal/mempool/error_test.go:74]: No reason for the prefix and to be consistent:
```Go
			t.Errorf("#%d: got: %s want: %s", i, result, test.want)
```

**Review by davecgh — CHANGES_REQUESTED** [review]
Round 1 of latest updates.

File: internal/mempool/error.go [hunk 7]
```diff
@@ -10,124 +10,252 @@ import (
...
+
+	// Rejected due to checkpoint.
+	case errors.Is(err, blockchain.ErrCheckpointTimeTooOld),
+		errors.Is(err, blockchain.ErrDifficultyTooLow),
+		errors.Is(err, blockchain.ErrBadCheckpoint),
+		errors.Is(err, blockchain.ErrForkTooOld):
+		code = wire.RejectCheckpoint
+		extracted = true
+
+	// Rejected due to insufficient feesq.
```
Comment (davecgh) [inline internal/mempool/error.go:211]: fees

File: blockmanager.go [hunk 8]
```diff
@@ -662,76 +662,63 @@ func (b *blockManager) handleDonePeerMsg(peer *peerpkg.Peer) {
...
 
-		// Rejected due to obsolete version.
-		case blockchain.ErrBlockVersionTooOld:
-			code = wire.RejectObsolete
+	// Rejected due to obsolete version.
+	case errors.Is(err, blockchain.ErrBlockVersionTooOld):
+		code = wire.RejectObsolete
+		reason = err.Error()
 
 		// Rejected due to checkpoint.
```
Comment (davecgh) [inline blockmanager.go:682]: Looks to be indented one too many.  Not gofmt?

File: internal/mempool/error_test.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2014 The btcsuite developers
+// Copyright (c) 2015-2020 The Decred developers
```
Comment (davecgh) [inline internal/mempool/error_test.go:2]: This is an entirely new file.  The copyright shouldn't be attributed to btcsuite or pre 2020.

```Go
// Copyright (c) 2020 The Decred developers
// Use of ...
```

File: internal/mempool/error.go [hunk 10]
```diff
@@ -10,124 +10,252 @@ import (
...
+
+	// ErrTooManyVotes indicates the number of vote double spends exceeds the
+	// maximum allowed.
+	ErrTooManyVotes = ErrorKind("ErrTooManyVotes")
+
+	// ErrDuplicateRevocation indicates the revocation already exists in the
+	// mempool.
+	ErrDuplicateRevocation = ErrorKind("ErrDuplicateRevocation")
+
+	// ErrOldVote indicates the tickets votes on an old block height lower than
```
Comment (davecgh) [inline internal/mempool/error.go:62]:
```suggestion
	// ErrOldVote indicates the tickets votes on a block height lower than
```

File: internal/mempool/error.go [hunk 11]
```diff
@@ -10,124 +10,252 @@ import (
...
+	// the minimum allowed by the mempool.
+	ErrOldVote = ErrorKind("ErrOldVote")
+
+	// ErrAlreadyExists indicates the transaction already exists on the
+	// main chain and is fully spent.
```
Comment (davecgh) [inline internal/mempool/error.go:67]:
```suggestion
	// main chain and is not fully spent.
```

File: internal/mempool/error.go [hunk 12]
```diff
@@ -10,124 +10,252 @@ import (
...
+
+	// ErrSeqLockUnmet indicates the transaction sequence locks are not active.
+	ErrSeqLockUnmet = ErrorKind("ErrSeqLockUnmet")
+
+	// ErrInsufficientPriority indicates the non-stake transaction has a
+	// priority lower than the minimum allowed.
+	ErrInsufficientPriority = ErrorKind("ErrInsufficientPriority")
+
+	// ErrFeeTooHigh indicates the transaction pays fees above the maximum
+	// allowed by the mempool.
```
Comment (davecgh) [inline internal/mempool/error.go:78]:
```suggestion
	// allowed by the active policy.
```

File: internal/mempool/error.go [hunk 13]
```diff
@@ -10,124 +10,252 @@ import (
...
+
+	// ErrTooManyTSpends indicates the number of tspend hashes exceeds
+	// the maximum allowed.
+	ErrTooManyTSpends = ErrorKind("ErrTooManyTSpends")
+
+	// ErrTSpendMinedOnAncestor indicates the referenced tspend was already
+	// mined on an ancestor block.
+	ErrTSpendMinedOnAncestor = ErrorKind("ErrTSpendMinedOnAncestor")
+
+	// ErrTSpendInvalidExpiry indicates the tspend expiry is invalid.
```
Comment (davecgh) [inline internal/mempool/error.go:92]:
```suggestion
	// ErrTSpendInvalidExpiry indicates the treasury spend expiry is invalid
	// or too far in the future according to the active policy.
```

**Review by davecgh — CHANGES_REQUESTED** [review]

File: blockmanager.go (hunk 8 above)
Comment (dnldd) [reply blockmanager.go:682]: using `goimports`, linter checks with `gofmts` however.

**Review by davecgh — APPROVED** [review]
